File: python_mid_june_2024/36. Valid Sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        rows = [set() for _ in range(9)]
        cols = [set() for _ in range(9)]
        cells = [set() for _ in range(9)]

        for i in range(9):
            for j in range(9):
                if board[i][j] == '.':
                    continue
                
                cell_idx = (i//3)*3 + j//3
                if board[i][j] in rows[i] or board[i][j] in cols[j] or board[i][j] in cells[cell_idx]:
                    return False

                rows[i].add(board[i][j])
                cols[j].add(board[i][j])
                cells[cell_idx].add(board[i][j])
        return True

File: java_july/36. Valid Sudoku.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        int[][] rows = new int[9][9];
        int[][] cols = new int[9][9];
        int[][] cells = new int[9][9];
        
        for (int i=0; i<9; i++) {
            for (int j=0; j<9; j++) {
                if (board[i][j] != '.') {
                    int num = Character.getNumericValue(board[i][j])-1;
                    int cellIdx = (i/3)*3 + j/3;
                    if (rows[i][num] == -1 || cols[j][num] == -1 || cells[cellIdx][num] == -1) {
                        return false;
                    }
                    else{
                        rows[i][num] = -1;
                        cols[j][num] = -1;
                        cells[cellIdx][num] = -1;
                    }
                }
            }
        }
        return true;
    }
}